ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.io.Closeable;
+import java.io.File;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Component that removes the "dead" data from the set of provided log segments and reclaims space.
+ * <p/>
+ * It does this by copying the valid data into swap spaces and atomically replacing the segments being compacted with
+ * the swap spaces.
+ */
+class BlobStoreCompactor {
+  static final String INDEX_SEGMENT_READ_JOB_NAME = "blob_store_compactor_index_segment_read";
+  static final String LOG_SEGMENT_COPY_JOB_NAME = "blob_store_compactor_log_segment_copy";
+  static final String TARGET_INDEX_CLEAN_SHUTDOWN_FILE_NAME = "compactor_clean_shutdown";
+  static final FilenameFilter TEMP_LOG_SEGMENTS_FILTER = new FilenameFilter() {
+    private final String SUFFIX = LogSegmentNameHelper.SUFFIX + TEMP_LOG_SEGMENT_NAME_SUFFIX;
+
+    @Override
+    public boolean accept(File dir, String name) {
+      return name.endsWith(SUFFIX);
+    }
+  };
+
+  private static final long WAIT_TIME_FOR_CLEANUP_MS = 5 * Time.MsPerSec;
+  private static final String TEMP_LOG_SEGMENT_NAME_SUFFIX = BlobStore.SEPARATOR + "temp";
```

Comment:
This is the separator for the different parts of the names of files of store components so file path separator won't be appropriate. I cannot think of a better name other than what is there for now.